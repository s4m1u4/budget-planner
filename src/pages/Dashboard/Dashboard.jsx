import React, { useEffect, useState } from "react";
import { calculateRecords } from "./helpers";
import { Charts, Overview, Records } from "./components";
import { CircularProgress, Container } from "@mui/material";

import {
  Grid,
  Wrapper,
  DashboardTitle,
  ProgressWrapper,
} from "./Dashboard.styles";

export const Dashboard = ({
  histories,
  getHistories,
  setNewHistory,
  categories,
  getCategories,
  setNewCategory,
}) => {
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      await getCategories();
      await getHistories();
      setIsLoading(false);
    };

    fetchData();
  }, [getCategories, getHistories]);

  const records = calculateRecords(histories, categories);

  const lastRecords = records.slice(0, 5);

  return (
    <Container>
      {isLoading ? (
        <ProgressWrapper>
          <CircularProgress />
        </ProgressWrapper>
      ) : (
        <Wrapper>
          <DashboardTitle>Dashboard page</DashboardTitle>
          <Grid>
            <Overview records={records} />
            <Records
              histories={histories}
              categories={categories}
              lastRecords={lastRecords}
              setNewHistory={setNewHistory}
            />
            <Charts
              records={records}
              histories={histories}
              categories={categories}
              setNewCategory={setNewCategory}
            />
          </Grid>
        </Wrapper>
      )}
    </Container>
  );
};
