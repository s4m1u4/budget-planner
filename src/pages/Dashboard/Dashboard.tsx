import React, { FC, useCallback, useEffect, useState } from "react";
import { calculateRecords } from "helpers";
import { Charts, Overview, Records } from "./components";
import { CircularProgress, Container } from "@mui/material";
import { DashboardProps } from "./types";
import { IRecord } from "types";

import {
  Grid,
  Wrapper,
  DashboardTitle,
  ProgressWrapper,
} from "./Dashboard.styles";

export const Dashboard: FC<DashboardProps> = ({
  histories,
  getHistories,
  setNewHistory,
  categories,
  getCategories,
  setNewCategory,
  deleteCategory,
}) => {
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const fetchData: () => Promise<void> = useCallback(async () => {
    setIsLoading(true);
    await getCategories();
    await getHistories(null, 10000);
    setIsLoading(false);
  }, [getCategories, getHistories]);

  useEffect(() => {
    fetchData();

    return () => {
      setIsLoading(false);
    };
  }, [fetchData]);

  const records: IRecord[] = calculateRecords(histories, categories);

  const lastRecords: IRecord[] = records.slice(-5).reverse();

  return (
    <Container>
      {isLoading ? (
        <ProgressWrapper>
          <CircularProgress />
        </ProgressWrapper>
      ) : (
        <Wrapper>
          <DashboardTitle>Dashboard page</DashboardTitle>
          <Grid>
            <Overview records={records} />
            <Records
              categories={categories}
              lastRecords={lastRecords}
              setNewHistory={setNewHistory}
            />
            <Charts
              records={records}
              categories={categories}
              deleteCategory={deleteCategory}
              setNewCategory={setNewCategory}
            />
          </Grid>
        </Wrapper>
      )}
    </Container>
  );
};
