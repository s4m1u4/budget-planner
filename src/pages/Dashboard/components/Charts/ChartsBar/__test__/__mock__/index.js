export { TITLE, DATA } from "./mock";
