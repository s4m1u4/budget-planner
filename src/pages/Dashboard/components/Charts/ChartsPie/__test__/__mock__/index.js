export { TITLE, DATA_INCOME } from "./mock";
