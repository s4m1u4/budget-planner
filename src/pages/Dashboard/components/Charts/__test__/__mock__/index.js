export { CATEGORIES, RECORDS } from "./mock";
