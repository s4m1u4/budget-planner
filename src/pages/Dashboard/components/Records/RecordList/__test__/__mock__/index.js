export { LAST_RECORDS } from "./mock";
