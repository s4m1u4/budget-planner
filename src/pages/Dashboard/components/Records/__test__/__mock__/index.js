export { CATEGORIES, LAST_RECORDS } from "./mock";
