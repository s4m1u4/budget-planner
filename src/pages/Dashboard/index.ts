export { DashboardStore } from "./DashboardStore";
export { DashboardContainer } from "./DashboardContainer";
