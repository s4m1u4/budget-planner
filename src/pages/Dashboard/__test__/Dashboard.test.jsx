import { Dashboard } from "../Dashboard";
import { act, render, screen } from "@testing-library/react";
import { CATEGORIES, HISTORIES } from "./__mock__";

const renderComponentAsync = async () => {
  await act(async () => {
    await render(
      <Dashboard
        histories={HISTORIES}
        categories={CATEGORIES}
        getHistories={getHistories}
        setNewHistory={setNewHistory}
        getCategories={getCategories}
        setNewCategory={setNewCategory}
        deleteCategory={deleteCategory}
      />
    );
  });
};

const getHistories = jest.fn();
const setNewHistory = jest.fn();
const getCategories = jest.fn();
const setNewCategory = jest.fn();
const deleteCategory = jest.fn();

describe("Dashboard page", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderComponentAsync();
    expect(baseElement).toMatchSnapshot();
  });

  describe("should render", () => {
    beforeEach(() => {
      renderComponentAsync();
    });

    it("a title", () => {
      expect(screen.getByText(/dashboard page/i)).toBeInTheDocument();
    });
  });
});
