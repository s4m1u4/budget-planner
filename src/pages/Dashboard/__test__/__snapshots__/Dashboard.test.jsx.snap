// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page snapshot matching 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div
        class="css-1is0qkf"
      >
        <h1
          class="css-15poin4"
        >
          Dashboard page
        </h1>
        <div
          class="css-1myhdov"
        >
          <div
            class="css-1a8z9k1"
          >
            <div
              class="css-13o2qpx"
            >
              <h3
                class="css-1xqtn9i"
              >
                Overview
              </h3>
            </div>
            <div
              class="css-1gk4ta0"
            >
              <div
                class="css-gg4vpm"
              >
                <p
                  class="css-15v0w87"
                >
                  Amount:
                </p>
                <span
                  class="css-15v0w87"
                >
                  $3,190.00
                </span>
              </div>
            </div>
          </div>
          <div
            class="css-sai3sf"
          >
            <div
              class="css-13o2qpx"
            >
              <h3
                class="css-1xqtn9i"
              >
                Last records
              </h3>
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
                data-modal="modal"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddCircleOutlineSharpIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="css-1gk4ta0"
            >
              <div
                class="css-7wb1st"
              >
                <div
                  class="css-gg4vpm"
                >
                  <p
                    class="css-0"
                  >
                    income
                     from 
                    Work
                  </p>
                  <span
                    class="css-81iyw9"
                    color="income"
                  >
                    +$836.00
                  </span>
                </div>
                <div
                  class="css-gg4vpm"
                >
                  <p
                    class="css-0"
                  >
                    income
                     from 
                    Work
                  </p>
                  <span
                    class="css-81iyw9"
                    color="income"
                  >
                    +$2,354.00
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-a5nrzb"
          >
            <div
              class="css-13o2qpx"
            >
              <h3
                class="css-1xqtn9i"
              >
                Charts
              </h3>
              <div
                class="css-i5q2k0"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
                  data-charts="charts-pie"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="PieChartIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 2v20c-5.07-.5-9-4.79-9-10s3.93-9.5 9-10zm2.03 0v8.99H22c-.47-4.74-4.24-8.52-8.97-8.99zm0 11.01V22c4.74-.47 8.5-4.25 8.97-8.99h-8.97z"
                    />
                  </svg>
                </button>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
                  data-charts="charts-bar"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="BarChartIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z"
                    />
                  </svg>
                </button>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
                  data-modal="modal-create"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="AddCircleOutlineSharpIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                    />
                  </svg>
                </button>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-npdkjk-MuiButtonBase-root-MuiIconButton-root"
                  data-modal="modal-delete"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="DeleteForeverRoundedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm3.17-6.41a.9959.9959 0 0 1 0-1.41c.39-.39 1.02-.39 1.41 0L12 12.59l1.41-1.41c.39-.39 1.02-.39 1.41 0s.39 1.02 0 1.41L13.41 14l1.41 1.41c.39.39.39 1.02 0 1.41s-1.02.39-1.41 0L12 15.41l-1.41 1.41c-.39.39-1.02.39-1.41 0a.9959.9959 0 0 1 0-1.41L10.59 14l-1.42-1.41zM18 4h-2.5l-.71-.71c-.18-.18-.44-.29-.7-.29H9.91c-.26 0-.52.11-.7.29L8.5 4H6c-.55 0-1 .45-1 1s.45 1 1 1h12c.55 0 1-.45 1-1s-.45-1-1-1z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="css-1gk4ta0"
            >
              <div
                class="css-ivrut9"
              >
                <div>
                  <h4
                    class="css-sqlmve"
                  >
                    Income
                  </h4>
                  <div
                    class="recharts-responsive-container"
                    style="width: 99%; height: 300px;"
                  />
                </div>
                <div>
                  <h4
                    class="css-sqlmve"
                  >
                    Expense
                  </h4>
                  <div
                    class="recharts-responsive-container"
                    style="width: 99%; height: 300px;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
